server/src/main/java/io/crate/execution/engine/indexing/IndexWriterProjector.java

```diff
@@ -95,34 +95,83 @@ public IndexWriterProjector(ClusterService clusterService,
-        ShardUpsertRequest.Builder builder = new ShardUpsertRequest.Builder(
-            txnCtx.sessionSettings(),
-            ShardingUpsertExecutor.BULK_REQUEST_TIMEOUT_SETTING.get(settings),
-            overwriteDuplicates ? DuplicateKeyAction.OVERWRITE : DuplicateKeyAction.UPDATE_OR_FAIL,
-            true,
-            null,
-            new Reference[]{rawSourceReference},
-            null,
-            jobId
-        );
 
-        ItemFactory<ShardUpsertRequest.Item> itemFactory = (id, pkValues, autoGeneratedTimestamp) -> ShardUpsertRequest.Item.forInsert(
-            id,
-            pkValues,
-            autoGeneratedTimestamp,
-            new Object[]{source.value()},
-            null
-        );
+        TimeValue timeout = ShardingUpsertExecutor.BULK_REQUEST_TIMEOUT_SETTING.get(settings);
+        boolean continueOnError = true;
+        DuplicateKeyAction duplicateKeyAction = overwriteDuplicates
+            ? DuplicateKeyAction.OVERWRITE
+            : DuplicateKeyAction.UPDATE_OR_FAIL;
+        String[] updateColumns = null;
+        Symbol[] returnValues = null;
+        ShardingUpsertExecutor.BULK_REQUEST_TIMEOUT_SETTING.get(settings);
+        final AtomicReference<Reference[]> targetColumns = new AtomicReference<>();
+        final SessionSettings sessionSettings = txnCtx.sessionSettings();
+        Function<ShardId, ShardUpsertRequest> createRequest = shardId -> {
+            Reference[] insertColumns = targetColumns.get();
+            assert insertColumns != null : "targetColumns must be set";
+            return new ShardUpsertRequest(
+                shardId,
+                jobId,
+                continueOnError,
+                duplicateKeyAction,
+                sessionSettings,
+                updateColumns,
+                insertColumns,
+                returnValues
+            ).timeout(timeout);
+        };
+        ItemFactory<ShardUpsertRequest.Item> itemFactory = (id, pkValues, autoGeneratedTimestamp) -> {
+            Map<String, Object> doc = docInput.value();
+            Set<String> keys = doc.keySet();
+            Reference[] references = new Reference[keys.size()];
+            int i = 0;
+            for (String key : doc.keySet()) {
+                ColumnIdent column = ColumnIdent.fromNameSafe(key, List.of());
+                Reference reference = table.getReference(column);
+                if (reference == null) {
+                    reference = table.getDynamic(column, true, sessionSettings.errorOnUnknownObjectKey());
+                }
+                references[i] = reference;
+                i++;
+            }
+            Reference[] prevReferences = targetColumns.get();
+            if (prevReferences != null && !Arrays.deepEquals(prevReferences, references)) {
+                targetColumns.set(null);
+                throw new BatchIterators.NewPartitionException();
+            }
+            targetColumns.set(references);
```

Comment:
@seut could you have a look at this approach? Specifically the approach with the exception for partitioning/chunking control.

It feels a bit hacky.
An alternative would be to rework how the `ShardingUpsertExecutor` and its components work or ditch changing the request payload altogether, stick to `_raw`/`_doc` and continue handle the indexing  on the shard (e.g. via a new `DocValueIndexer`, which we may be able to use for translog replay as well?)